### PR TITLE
[13.0][MIG] sale_exception

### PR DESCRIPTION
**Pull request by jaredkipe**
Requires https://github.com/OCA/server-tools/pull/1760

Cleanup per docs including squashing "UPD" commits.
Unit tests run and pass.

### CODE REVIEW

**Comment by jaredkipe** [comment]
Requires https://github.com/OCA/server-tools/pull/1760

File: sale_exception/wizard/sale_exception_confirm.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Copyright 2011 Akretion, Sodexis
+# Copyright 2018 Akretion
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import api, fields, models
+
+
+class SaleExceptionConfirm(models.TransientModel):
+    _name = 'sale.exception.confirm'
```
Comment (HviorForgeFlow) [inline sale_exception/wizard/sale_exception_confirm.py]:
```suggestion
    _name = 'sale.exception.confirm'
    _description = "Sale exception confirm Wizard"
```

**Review by HviorForgeFlow — CHANGES_REQUESTED** [review]
You may also need to rebase 13.0 branch to apply new pre-commit changes.
Please split Migration commit into pre-commit + migration commit as you can see [here](https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-13.0#technical-method-to-migrate-a-module-from-120-to-130-branch)

File: sale_exception/models/sale.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+
+class SaleOrder(models.Model):
+    _inherit = ['sale.order', 'base.exception']
+    _name = 'sale.order'
+    _order = 'main_exception_id asc, date_order desc, name desc'
+
+    @api.model
+    def _exception_rule_eval_context(self, rec):
+        res = super(SaleOrder, self)._exception_rule_eval_context(rec)
```
Comment (mmequignon) [inline sale_exception/models/sale.py:28]: As it is python3, you can use `res = super()._exception_rule_eval_context(rec)` instead.

File: sale_exception/models/sale.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        res['sale'] = rec
+        return res
+
+    @api.model
+    def _reverse_field(self):
+        return 'sale_ids'
+
+    def detect_exceptions(self):
+        all_exceptions = super(SaleOrder, self).detect_exceptions()
```
Comment (mmequignon) [inline sale_exception/models/sale.py:37]: Same here

File: sale_exception/models/sale.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        order_set = self.search([('state', '=', 'draft')])
+        order_set.detect_exceptions()
+        return True
+
+    def _fields_trigger_check_exception(self):
+        return ['ignore_exception', 'order_line', 'state']
+
+    @api.model
+    def create(self, vals):
+        record = super(SaleOrder, self).create(vals)
```
Comment (mmequignon) [inline sale_exception/models/sale.py:53]: Etc…

File: sale_exception/models/sale_order_line.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+    def _get_main_records(self):
+        return self.mapped('order_id')
+
+    @api.model
+    def _reverse_field(self):
+        return 'sale_ids'
+
+    def _detect_exceptions(self, rule):
+        records = super(SaleOrderLine, self)._detect_exceptions(rule)
```
Comment (mmequignon) [inline sale_exception/models/sale_order_line.py:23]: Same in this file.

**Review by mmequignon — CHANGES_REQUESTED** [review]
Looks good !
A few remarks however, that will simplify future migrations.

File: sale_exception/wizard/sale_exception_confirm.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# Copyright 2011 Akretion, Sodexis
+# Copyright 2018 Akretion
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import api, fields, models
```
Comment (sebalix) [inline sale_exception/wizard/sale_exception_confirm.py]: `api` is not used (lint fails on that line)

**Comment by rousseldenis** [comment]
> Requires #1760

https://github.com/OCA/server-tools/pull/1760

**Comment by rousseldenis** [comment]
@jaredkipe Could you please read https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-13.0#technical-method-to-migrate-a-module-from-120-to-130-branch ?

And in particular things about pre-commit ?

Thanks for the migration of this great module!

**Comment by leemannd** [comment]
Not tested yet, but I'm wondering one point. What happens with the following configuration:

- SO created automatically by an external system
- 20 SO Lines or more
- 2 kinds of exceptions triggered by an SOLine
- More than 2 SOLines are in error

This would be a nightmare to fix the SO by an end-user. I know this is not the purpose of migrations, but it would be a nice improvement to maybe:

- display the lines that are in error, with the error type in the wizard
- log a note with the previous message

What do you think?

**Comment by jaredkipe** [comment]
I will work on this when the base module is merged.

**Comment by guewen** [comment]
@leemannd this may be what you are looking for: https://github.com/OCA/sale-workflow/pull/1194
However, we need to get this merged before. @jaredkipe, I ran pre-commit here: https://github.com/hibou-io/oca-sale-workflow/pull/1 but you may need to rebase your branch (same thing for https://github.com/OCA/server-tools/pull/1760).

**Comment by leemannd** [comment]
@guewen Absolutely! :partying_face:

File: sale_exception/models/sale_order_line.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        return records.mapped('order_id')
+
+    @api.model
+    def _exception_rule_eval_context(self, rec):
+        # We keep this only for backward compatibility, because some existing
+        # rules may use the variable "sale_line". But we should remove this
+        # code during v13 migration. The record is already available in obj and
```
Comment (tfossoul) [inline sale_exception/models/sale_order_line.py:30]: What about this ?

**Comment by simahawk** [comment]
Replaced by https://github.com/OCA/sale-workflow/pull/1376 to speed it up.

**Pull request closed without merging**